### ISSUE TITLE
Transformer CONT1 may be Float64 when COD1=1; you declare it invalid

### ISSUE THREAD

**Issue by byronbest** [lanl-ansi/PowerModels.jl#459]
Quoting the PTIv33 documentation:
CONT1, CONT2, CONT3:

**The bus for which voltage is to be controlled by the transformer turns ratio adjustment 
option when CODX is 1.** 

Trying to load example data from GridPACK gets
```
Loading IEEE14_PTIv33.raw

[error | PowerModels]: value '1.50000' for CONT1 in section TRANSFORMER TWO WINDING is not of type Int64.
[error | PowerModels]: Parsing failed at line 57: value '1.50000' for CONT1 in section TRANSFORMER TWO WINDING is not of type Int64.
```

**Comment by pseudocubic** [comment]
According to the manual, CONT1 is "The bus number, or extended bus name enclosed in single quotes", and can therefore never be a float. 

I investigated the files in gridpack, in "IEEE14_PTIv33.raw", which is the file I assume you are referring to, there is a missing field at the end of each transformer line 1, VECGRP, which I assume was added in v33.3, and therefore this file is probably between v33.0-v33.2, and isn't explicitly supported by this parser. The parser thinks that the value meant to be RMA1 is actually CONT1 because of this missing field.

I have an idea about how to support this, but it will require some refactoring of the current parse algorithm, and isn't a big priority for us right now. I will keep this issue open and try to address it in the future.

**Comment by byronbest** [comment]
Thanks, David!
I have a julia parser that reads into DataFrames, and handles missing
fields at the end of lines by setting them `missing`.
I split the file into sections and process each block.
It's much smaller and less prone to errors, compared to reading the file
line by line, splitting etc. which has it's roots in the Fortran and C++.

You get an Array of Vectors instead of Dict of Sets, so I'm currently
looking deeper to see how much pain that creates. Hoping ordered Vectors
will be a boon rather than a bane.

Best,
Byron

**Comment by ccoffrin** [comment]
It looks like we should be more specific about which sub-version of PSSE v33 we are supporting.

As for supporting a wider range, as it is convenient, we should try to make the parser robust to messy inputs.  However, robust support for a wide range of file variants is probably better left to PowerSystems.jl

**Comment by ccoffrin** [comment]
@byronbest, I think this feature add (https://github.com/lanl-ansi/PowerModels.jl/pull/465) has addressed this issue.  Can you check if the `master` branch will parse your data file?

**Comment by byronbest** [comment]
Will do-- note that I was trying to load the example data files included
with GridPACK, so you could add them to the testing, too.

Documentation at AESO implies that the CONT1 field could be a ratio or an
ID integer. I was quick to first blame GridPACK, then wrote my own version
in Julia that loads into DataFrame and sets missing to unknown.

**Comment by byronbest** [comment]
BAD News, sorry ... I wiped my .julia to be sure
(v1.1) pkg> dev PowerModels

Loading IEEE14_PTIv33.raw
[info | PowerModels]: The PSS(R)E parser currently supports buses, loads,
shunts, generators, branches, transformers, and dc lines
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO,
EVHI, EVLO
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[error | PowerModels]: value '1.50000' for CONT1 in section TRANSFORMER TWO
WINDING is not of type Int64.
[error | PowerModels]: Parsing failed at line 57: value '1.50000' for CONT1
in section TRANSFORMER TWO WINDING is not of type Int64.
ERROR: Parsing failed at line 57: value '1.50000' for CONT1 in section
TRANSFORMER TWO WINDING is not of type Int64.

**Comment by pseudocubic** [comment]
I have double checked the PSSE v33.3 documentation, and CONT is not a ratio, but controls a ratio that is applied to a calculation internally. It is either a bus number, and therefore an Integer, or an extended bus name (therefore should be surrounded by single quotes), with the caveat that there can be a minus sign to indicate a different ratio adjustment. The number that is getting parsed wrong in that file is RMA1, because there is no VECGRP on the previous line. However, in PowerModels master I have included a fix for missing (default) values (see PR #465 ) that parses each line of a transformer entry separately.

On PowerModels master this is the result I get when I load the file `GridPACK-3.2/src/applications/data_sets/raw/IEEE14_PTIv33.raw`:

```
julia> using PowerModels

julia> pti = PowerModels.parse_file("IEEE14_PTIv33.raw")
[info | PowerModels]: The PSS(R)E parser currently supports buses, loads, shunts, generators, branches, transformers, and dc lines
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in BUS are missing: NVHI, NVLO, EVHI, EVLO
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in LOAD are missing: INTRPT
[warn | PowerModels]: The following fields in TRANSFORMER are missing: VECGRP
[warn | PowerModels]: The following fields in TRANSFORMER TWO-WINDING LINE 2 are missing: CNXA1
[warn | PowerModels]: The following fields in TRANSFORMER are missing: VECGRP
[warn | PowerModels]: The following fields in TRANSFORMER TWO-WINDING LINE 2 are missing: CNXA1
[warn | PowerModels]: The following fields in TRANSFORMER are missing: VECGRP
[warn | PowerModels]: The following fields in TRANSFORMER TWO-WINDING LINE 2 are missing: CNXA1
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 4 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 1 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 12 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 20 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 2 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 6 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 11 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 13 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 5 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 15 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 16 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 14 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 7 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 8 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 17 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 10 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 19 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 9 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 18 to +/- 60.0 deg.
[warn | PowerModels]: angmin and angmax values are 0, widening these values on branch 3 to +/- 60.0 deg.
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 4 to 651.7693
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 1 to 1943.3753
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 12 to 1099.9023
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 20 to 480.1216
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 2 to 527.2551
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 6 to 658.7239
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 11 to 686.916
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 13 to 1340.1421
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 5 to 661.3166
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 15 to 579.3328
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 16 to 406.1471
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 14 to 404.9974
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 7 to 2739.0808
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 8 to 548.9677
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 17 to 312.073
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 10 to 828.1844
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 19 to 217.5559
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 9 to 426.3491
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 18 to 578.6164
[warn | PowerModels]: this code only supports positive rate_a values, changing the value on branch 3 to 594.6825
Dict{String,Any} with 12 entries:
  "bus"            => Dict{String,Any}("4"=>Dict{String,Any}("zone"=>2,"bus_i"=>4,"bus_type"=>1,"…
  "source_type"    => "pti"
  "name"           => "ieee14_ptiv33"
  "dcline"         => Dict{String,Any}()
  "source_version" => v"33.0.0"
  "gen"            => Dict{String,Any}("4"=>Dict{String,Any}("pg"=>0.0,"model"=>2,"shutdown"=>0.0…
  "branch"         => Dict{String,Any}("4"=>Dict{String,Any}("br_r"=>0.05811,"rate_a"=>6.51769,"s…
  "storage"        => Dict{String,Any}()
  "baseMVA"        => 100.0
  "per_unit"       => true
  "shunt"          => Dict{String,Any}("1"=>Dict{String,Any}("source_id"=>Any[9, " 1"],"shunt_bus…
  "load"           => Dict{String,Any}("4"=>Dict{String,Any}("source_id"=>Any[5, "1 "],"load_bus"…
```

The only thing I can think of is somehow PowerModels got pinned to an earlier version in your environment and isn't pulling the latest master. I have confirmed that my local copy of PowerModels is up-to-date with this master.

**Comment by byronbest** [comment]
more mystery, thought this was same as dev PowerModels, and it acts so
I had thought this was the master, maybe not?

(v1.1) pkg> dev https://github.com/lanl-ansi/PowerModels.jl.git

and I get the same old broken at the missing field and next line confusion

**Comment by pseudocubic** [comment]
@byronbest can you provide the output of `git describe --tags` for PowerModels (it should be `v0.9.3-14-g0bf198a`), and `] st` inside julia? Also, can you confirm we are using the same file from GridPack by checking the sha1 (mine is `a1d7f86496872540d8df76a4ff8d52dea83d86d6`)

**Comment by byronbest** [comment]
Found something -- v0.9.3-1 -- after rm, dev, and update

bbest@stockholm-ubuntu:~/.julia/dev/PowerModels$ git describe --tags
v0.9.3-1-gb08d925


/home/bbest/.julia/environments/v1.1/Manifest.toml:    PowerModels =
"c36e90e8-916a-50a6-bd94-075b64ef4655"
/home/bbest/.julia/environments/v1.1/Project.toml:PowerModels =
"c36e90e8-916a-50a6-bd94-075b64ef4655"
::::::::::::::
registries/General/P/PowerModels/Package.toml
::::::::::::::
name = "PowerModels"
uuid = "c36e90e8-916a-50a6-bd94-075b64ef4655"
repo = "https://github.com/lanl-ansi/PowerModels.jl.git"
::::::::::::::
registries/General/P/PowerModels/Versions.toml
::::::::::::::
...
["0.9.3"]
git-tree-sha1 = "1424521b04d5a2e4b6f084cafeaac6ab905bc9f2"

(v1.1) pkg> st
    Status `~/.julia/environments/v1.1/Project.toml`
  [c52e3926] Atom v0.7.14
  [336ed68f] CSV v0.4.3
  [a93c6f00] DataFrames v0.17.1+ [`~/.julia/dev/DataFrames`]
  [1313f7d8] DataFramesMeta v0.4.1
  [f67ccb44] HDF5 v0.11.0
  [b6b21f68] Ipopt v0.5.4+ [`~/.julia/dev/Ipopt`]
  [4076af6c] JuMP v0.18.5
  [e5e0dc1b] Juno v0.5.4
  [c36e90e8] PowerModels v0.9.3+ [`~/.julia/dev/PowerModels`]
  [e690365d] PowerSimulations v0.1.0 [`~/.julia/dev/PowerSimulations`]
  [c512b964] PowerSystems v0.2.1 [`~/.julia/dev/PowerSystems`]
  [8bb1440f] DelimitedFiles

bbest@stockholm-ubuntu:/datapool/wecc/powerflow$ sha1sum ds/*
1461e06ba6db32105800bd31c277e19b9fa00020  ds/118_PTIv33.raw
c0d903762c5e3adc8faf3c7e153e9c8befd58a2f  ds/14.raw
f04551d53d0226ed0b88d534612df9772c6fcf55  ds/EuropeanOpenModel_v33.raw
a1d7f86496872540d8df76a4ff8d52dea83d86d6  ds/IEEE14_PTIv33.raw
58081a52516221601ca39a0c6a867ee1c50474a9  ds/Polish_model_v33.raw

I also cannot read Euro nor Polish

**Comment by pseudocubic** [comment]
Definitely there is a problem with your local copy of the PowerModels repo. One thing I've noticed is that inside julia `rm PowerModels` doesn't actually get rid of the folder inside `.julia/dev`, so even when you `rm` and `dev`, it just goes back to that same directory.

First thing, is your local repo pointing to the right remote? `git remote get-url origin` should result in `https://github.com/lanl-ansi/PowerModels.jl.git`, if its wrong you can update it with `git remote set-url origin https://github.com/lanl-ansi/PowerModels.jl.git`

You can then try to update from the repo directory with git: `git fetch origin && git pull origin master`. If this fails then your local copy of the repo probably has changes made to it (which would prevent `] up` inside julia from doing anything), and needs to be reset to remote (`git fetch origin && git reset --hard origin/master`)

**Comment by pseudocubic** [comment]
I have also confirmed that with the current revision of PowerModels all of the files you listed can be parsed (except `14.raw`, which I don't seem to have in the GridPack-3.2 clone).

**Comment by byronbest** [comment]
I always start a report  by rm -rf ~/.julia then make it fail with known
install.
Thanks for your help

Maybe git cache somewhere here is returning ... hey! now it's right!

Must be matching 0.9.3+ and giving up?

bbest@stockholm-ubuntu:~/.julia/dev/PowerModels$ git pull origin master
From https://github.com/lanl-ansi/PowerModels.jl
 * branch            master     -> FETCH_HEAD
Already up to date.
bbest@stockholm-ubuntu:~/.julia/dev/PowerModels$ git describe --tags
v0.9.3-14-g0bf198a

(v1.1) pkg> dev PowerModels
[ Info: Path `/home/bbest/.julia/dev/PowerModels` exists and looks like the
correct package, using existing path
 Resolving package versions...
  Updating `~/.julia/environments/v1.1/Project.toml`
  [c36e90e8] + PowerModels v0.9.3+ [`~/.julia/dev/PowerModels`]


And I can read the files! Thanks for your help!

**Comment by pseudocubic** [comment]
Glad you got it to work! 

Closing this issue, addressed by PR #465
